Optimize memory consumption during backup to S3

vitlibar opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimize memory consumption during backup to S3: files to S3 now will be copied directly without using `WriteBufferFromS3` (which could use a lot of memory).

This PR fixes https://github.com/ClickHouse/ClickHouse/issues/43083

vitlibar committed:
Copy files to S3 during backup directly without using WriteBufferFromS3 to decrease memory consumption.

CheSema commented on src/IO/S3/copyDataToS3.cpp:
```diff
@@ -0,0 +1,682 @@
+            auto outcome = client_ptr->UploadPart(req);
+            if (!outcome.IsSuccess())
+            {
+                abortMultipartUpload();
```
> However, if any part uploads are currently in progress, those part uploads might or might not succeed. As a result, it might be necessary to abort a given multipart upload multiple times in order to completely free all storage consumed by all parts.

As I see, this call is concurrent with other UploadPart calls. In that case it might be useless.

vitlibar replied:
I think I have to check the result of `aboutMultipartUpload` - and if it didn't succeed then I need to call `aboutMultipartUpload` again in the end. I'll change the code.

vitlibar replied:
done

CheSema commented on src/IO/StdStreamBufFromReadBuffer.cpp:
```diff
@@ -0,0 +1,111 @@
+    if (dir == std::ios_base::cur)
+        return seekpos(getCurrentPosition() + off, which);
+
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Wrong seek's base {}", static_cast<int>(dir));
```
Interesting. There is should be case for std::ios_base::end also.
I definitely have seen that S3 do '.seekg(0, body->end);'.
When it sets context-length in httpRequest (actually that seems like a bug because that information has been provided).

vitlibar replied:
The problem is that `ReadBuffer` doesn't have an interface to get the total size. However it's not a problem, I can pass it via a new parameter of the constructor of `StdStreamBufFromReadBuffer`.

vitlibar replied:
done

CheSema commented:
I like that approach. Actually this is the best way when we upload file from local disk to S3. 
I have several question. Mostly just to make history, what and why this code do.

When you copy file you know its size. Why do you still use exponential part upload size instead some fixed?

When you copy from S3 to S3 why do you prefer seakable reading from S3 instead UploadPartCopyRequest? My concern is the following. With seakable reading you read a file twice. First is to calc sha hash. Second is to upload it to S3. Indeed this read is made with fixed buffer size, memory consumption is low. However I think it is worth to do with existing code in order not to have one more realisation with UploadPartCopyRequest.

vitlibar commented:
> When you copy file you know its size. Why do you still use exponential part upload size instead some fixed?

Yes you're right, that code was copied from `WriteBufferFromS3` but there are no reason why we should keep that exponential approach for those copy functions.

> When you copy from S3 to S3 why do you prefer seakable reading from S3 instead UploadPartCopyRequest? My concern is the following. With seakable reading you read a file twice. First is to calc sha hash. Second is to upload it to S3. Indeed this read is made with fixed buffer size, memory consumption is low. However I think it is worth to do with existing code in order not to have one more realisation with UploadPartCopyRequest.

I don't quite understand. I don't prefer seekable reading over `UploadPartCopyRequest`. Copying from S3 to S3 with ``UploadPartCopyRequest` must always be better. Seekable reading is not for copying from S3 to S3, it's mostly for uploading data from local disks to S3.

CheSema commented:
> I don't quite understand. I don't prefer seekable reading over UploadPartCopyRequest

Sorry. Indeed, you don't. You have dedicated realisation for S3toS3 with UploadPartCopyRequest.

vitlibar committed:
Split upload into parts of the same size for smooth uploading.
Correctly use AbortMultipleUpload request.
Support std::ios_base::end StdStreamBufFromReadBuffer::seekpos().

vitlibar commented:
> When you copy file you know its size. Why do you still use exponential part upload size instead some fixed?

I did that. Now an uploaded file is splitted into parts of the same size.

vitlibar committed:
Fix build.

CheSema commented on src/IO/S3/copyDataToS3.cpp:
```diff
@@ -0,0 +1,723 @@
+                {
+                    task->req = fillUploadPartRequest(part_number, part_offset, part_size);
+
+                    schedule([this, task, task_finish_notify]()
```
Help me understand here, how many  concurrent uploads are run here? Is it limited thread pool?
My concern is that it runs a lot of tasks, each makes a new connection. It could run out of available inodes or face other limitation inside CH.

vitlibar replied:
It's normally used with `IOThreadPool` . It's limited of course. By default it runs up to 100 threads, so maximum 100 parallel uploads. But it actually must be smaller because `IOThreadPool` is also used for other tasks.

vitlibar replied:
In fact it should use its own thread pools `backups_thread_pool` and `restores_thread_pool`, so I changed the code to use the correct thread pools. Those thread pools are created specially for backups and restores.

CheSema commented on src/IO/StdStreamBufFromReadBuffer.cpp:
```diff
@@ -0,0 +1,111 @@
+    else if (dir == std::ios_base::cur)
+        return seekpos(getCurrentPosition() + off, which);
+    else if (dir == std::ios_base::end)
+        return seekpos(size, which);
```
You are ignoring arg off in that case. It might be wrong.
S3 sdk always do call seekg(0, end). However it is legal to call for example seekg(-10, end).

vitlibar replied:
Done, I made it correct

CheSema commented on src/IO/S3/copyDataToS3.cpp:
```diff
@@ -0,0 +1,723 @@
+
+            for (size_t part_number = 1; position < end_position; ++part_number)
+            {
+                if (multipart_upload_aborted)
```
Isn't it better to check this right before processUploadPartRequest? In order to stop work as fast as possible if non retryable error occurs. 
My concern is that the code have to wait all task which has been scheduled. If 1000 task were scheduled and only 100 of then is running and finishing with connection timeout. It would take 10 x timeout to finish the upload with an error.

CheSema replied:
It seems like the queue size is pretty big
>         auto queue_size = config.getUInt(".threadpool_writer_queue_size", 1000000);
and all task are about to be pushed to the schedule's queue. In case if first task fail no need to run the rest.

vitlibar replied:
> Isn't it better to also check multipart_upload_aborted right before processUploadPartRequest?

Done, I changed the code.

vitlibar committed:
Fix using std::ios_base::end in StdStreamFromReadBuffer::seekg().

vitlibar committed:
Abort multipart upload faster.

vitlibar commented:
The test `test_backup_restore_on_cluster/test_disallow_concurrency.py::test_concurrent_restores_on_different_node` is [flaky](https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZIb3VyKGNoZWNrX3N0YXJ0X3RpbWUpIGFzIGQsCmNvdW50KCksICBncm91cFVuaXFBcnJheShwdWxsX3JlcXVlc3RfbnVtYmVyKSwgIGFueShyZXBvcnRfdXJsKQpmcm9tIGNoZWNrcyB3aGVyZSAnMjAyMi0wNi0wMScgPD0gY2hlY2tfc3RhcnRfdGltZSBhbmQgdGVzdF9uYW1lIGxpa2UgJyV0ZXN0X2JhY2t1cF9yZXN0b3JlX29uX2NsdXN0ZXIvdGVzdF9kaXNhbGxvd19jb25jdXJyZW5jeS5weTo6dGVzdF9jb25jdXJyZW50X3Jlc3RvcmVzX29uX2RpZmZlcmVudF9ub2RlJScgYW5kIHRlc3Rfc3RhdHVzIGluICgnRkFJTCcsICdGTEFLWScpIGdyb3VwIGJ5IGQgb3JkZXIgYnkgZCBkZXNj)

Lloyd-Pottiger commented on src/IO/StdStreamFromReadBuffer.h:
```diff
@@ -0,0 +1,39 @@
+{
+public:
+    using Base = std::iostream;
+    StdStreamFromReadBuffer(std::unique_ptr<ReadBuffer> buf, size_t size) : Base(&stream_buf), stream_buf(std::move(buf), size) { }
+    StdStreamFromReadBuffer(ReadBuffer & buf, size_t size) : Base(&stream_buf), stream_buf(buf, size) { }
```
I have some questions about this, 

`Base(&stream_buf)` will call the following constructor

```cpp
template <class _CharT, class _Traits>
basic_streambuf<_CharT, _Traits>::basic_streambuf()
    : __binp_(nullptr),
      __ninp_(nullptr),
      __einp_(nullptr),
      __bout_(nullptr),
      __nout_(nullptr),
      __eout_(nullptr)
{
}

template <class _CharT, class _Traits>
basic_streambuf<_CharT, _Traits>::basic_streambuf(const basic_streambuf& __sb)
    : __loc_(__sb.__loc_),
      __binp_(__sb.__binp_),
      __ninp_(__sb.__ninp_),
      __einp_(__sb.__einp_),
      __bout_(__sb.__bout_),
      __nout_(__sb.__nout_),
      __eout_(__sb.__eout_)
{
}
```

so, `__ninp_` and `__einp_` are nullptr. But std::iostream will call `sgetc()` to get char, so we will always call `underflow()` which does not advance the position.

```cpp
inline _LIBCPP_HIDE_FROM_ABI_AFTER_V1
int_type sgetc() {
    if (__ninp_ == __einp_)
        return underflow();
    return traits_type::to_int_type(*__ninp_);
}
```
